lib/puppet/transaction.rb

```diff
@@ -229,7 +231,7 @@ def prefetch_if_necessary(resource)
   def apply(resource, ancestor = nil)
     status = resource_harness.evaluate(resource)
     add_resource_status(status)
-    event_manager.queue_events(ancestor || resource, status.events) unless status.failed?
+    event_manager.queue_events(ancestor || resource, status.events) unless status.failed? || status.failed_to_restart?
```

Comment:
this isn't necessary to the logic and its not your PR that makes this code hard to follow to be sure, but just for clarifying the flow could you add parens around the || conditional after unless? Or even just split it into multiple lines ie
```ruby
if !(status.failed? || status.failed_to_restart?)
  event_manager.queue_events(ancestor || resource, status.events)
end
```

Also the inline `||` in the param is a little too clever to follow IMO too...
what if we did something like:
```
ancestor ||= resource
event_manager.queue_events(ancestor, status.events)
...
```
Defer to your thoughts on it.